network/Message.cpp

```diff
@@ -743,15 +825,33 @@ void ExtractErrorMessageData(const Message& msg, int& player_id, std::string& pr
     }
 }
 
-void ExtractHostMPGameMessageData(const Message& msg, std::string& host_player_name, std::string& client_version_string) {
+void ExtractHostMPGameMessageData(const Message& msg, std::string& host_player_name, std::string& client_version_string,
+                                  bool& use_compression) {
     try {
         std::istringstream is(msg.Text());
         freeorion_xml_iarchive ia(is);
         ia >> BOOST_SERIALIZATION_NVP(host_player_name)
-           >> BOOST_SERIALIZATION_NVP(client_version_string);
+           >> BOOST_SERIALIZATION_NVP(client_version_string)
```

Comment:
should separately handle the new field so that, hopefully, it not being there doesn't prevent the other data from being used.

Reply:
Would using a caught exception be acceptable as a way of handling this? Something like:
```
// read everything else

try {
    ia >> BOOST_SERIALIZATION_NVP(use_compression);
} catch (const std::exception& dne) {
    use_compression = false;
}

// continue
```
I don't know if there are "better" ways to discover whether an element exists inside the `xml_iarchive`. Using an XML parser seems like overkill, and a substring search seems unreliable.

Reply:
I'm unsure if it will work, but I would guess that it should

Reply:
After some experimentation, I regret to report that it doesn't work.

```
try {
    std::istringstream is(msg.Text());
    freeorion_xml_iarchive ia(is);

    try {
        ia >> BOOST_SERIALIZATION_NVP(use_compression);
    } catch (const boost::archive::archive_exception& err) {
        use_compression = false;
    }
    // When ia goes out of scope, another archive_exception is thrown
} catch (const boost::archive::archive_exception& last_resort) {
    // Doesn't catch the exception
}
```
The `archive_exception` thrown when the tag is not found is caught and handled successfully. But then another `archive_exception` is thrown when `ia` goes out of scope, and that exception seems to be impossible to catch. 

It looks like an existing Boost issue: https://github.com/boostorg/serialization/issues/109. 

The only thing I can think of now is to create a new message type that includes the `use_compression` property. This would also be applied to the other messages that have been extended to communicate `use_compression`.

Reply:
Based on that Boost issue, put the `freeorion_xml_iarchive ia(is);` within the inner `try` (and get rid of the outer `try`), so that the archive is destructed immediately when the exception is thrown, rather than persisting into the immediately following `catch`. Doing this lets me catch such an exception without immediately aborting.

Reply:
Maybe we should allow to break compatibility between versions?

Reply:
I'd rather not if it can be avoided

Reply:
> Based on that Boost issue, put the `freeorion_xml_iarchive ia(is);` within the inner `try` (and get rid of the outer `try`), so that the archive is destructed immediately when the exception is thrown, rather than persisting into the immediately following `catch`. Doing this lets me catch such an exception without immediately aborting.

Collapsing the exception handling into a single `try` let's it run.
```
void ExtractHostMPGameMessageData(const Message& msg, std::string& host_player_name, std::string& client_version_string,
                                  bool& use_compression) {
    try {
        std::istringstream is(msg.Text());
        freeorion_xml_iarchive ia(is);
        ia >> BOOST_SERIALIZATION_NVP(host_player_name)
           >> BOOST_SERIALIZATION_NVP(client_version_string)
           >> BOOST_SERIALIZATION_NVP(use_compression);
    } catch (const boost::archive::archive_exception& err) {
        use_compression = false;
    } catch (const std::exception& err) {
        ErrorLogger() << "ExtractHostMPGameMessageData(const Message& msg, std::string& host_player_name, std::string& client_version_string, bool& use_compression) failed!  Message:\n"
                      << msg.Text() << "\n"
                      << "Error: " << err.what();
        throw err;
    }
}
```
But it also makes it impossible to tell whether it was `BOOST_SERIALIZATION_NVP(use_compression)` causing the `archive_exception`, which doesn't seem good.

It also makes `Message::ExtractJoinGameMessageData()` tricky since it also uses exceptions to detect the existence/validity of a `dependencies` element in the archive. Given the difficulties discovered above, the implementation of that function may already be broken.

I will try implementing a new message type to communicate compression use. Hopefully that side-steps all of this craziness.

Reply:
I'd be slightly surprised if there was just one case where a try/catch block is used to handle message deserialization like this that don't actually work, since I/we wasn't aware of the archive destructor throwing issue and probably added a few intended-to-be-optional message fields over time.

Reply:
https://github.com/freeorion/freeorion/commit/e937091c4540e765549e9147757117dfbe3f4bcf re-implements the transmission of compression usage requests to use a new message type `USE_COMPRESSION`. The changes to the serialization/deserialization of the existing messages are reverted.